Library/Homebrew/command.rb

```diff
@@ -0,0 +1,157 @@
+module Homebrew
+  class Command
+    attr_reader :command_name, :valid_options, :description, :help_output,
+      :man_output
+
+    def initialize_variables
+      @valid_options = []
+      @root_options = []
+      @argv = ARGV.dup
+    end
+
+    def options(&block)
+      initialize_variables
+      @parent = nil
+      instance_eval(&block)
+      generate_help_and_manpage_output
+    end
+
+    def cmd_name(cmd)
+      @command_name = cmd
+    end
+
+    def add_valid_option(option_hash)
+      @valid_options.push(option_hash)
+    end
+
+    def option(*args, **option_hash, &block)
+      option_hash[:option] = "--#{args[0]}"
+      option_name = option_hash[:option]
+      if @parent.nil?
+        @root_options.push(option_name)
+      else
+        hash = @valid_options.find { |x| x[:option] == @parent }
+        if hash[:child_options].nil?
+          hash[:child_options] = [option_name]
+        else
+          hash[:child_options].push(option_name)
+        end
+      end
+      add_valid_option(option_hash)
+      return unless block_given?
+      old_parent = @parent
+      @parent = option_name
+      instance_eval(&block)
+      @parent = old_parent
+    end
+
+    def desc(desc)
+      if @description.nil?
+        @description = desc
+      else
+        @description = <<-EOS.undent
+          #{@description}
+          #{desc}
+        EOS
+                             .strip
+      end
+    end
+
+    def argv_invalid_options_passed(argv)
+      argv_options_only = argv.select { |arg| /^--/ =~ arg }
+      argv_options_only = argv_options_only.uniq
+      valid_option_names =
+        @valid_options
+        .map { |option_hash| option_hash[:option] }
+
+      argv_options_only
+        .reject { |opt| valid_option_names.include?(opt.split("=", 2)[0]) }
+        .map { |opt| opt.split("=", 2)[0] }
```

Comment:
@MikeMcQuaid, so there are 2 types of options:

(1) options that expect a value for e.g. `--prune=<days>` or `--prune <days>` in `brew cleanup`.
(2) options that do not expect a value for e.g. `--quiet` in `brew command`. 

This function just checks whether the user has provided the correct option name or not. For e.g. `--prune1=20` instead of `--prune=20` or `--quiet1` instead of `--quiet`